### PR TITLE
BM 8 Likes & Dislikes

### PR DESCRIPTION
**Pull request by umair-khalid-brainx**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -36,3 +36,5 @@
 /node_modules
 
 .idea
+
+!/app/assets/images/
```
Comment (usama-javaid-brainx) [inline .gitignore]: why do we need to add assets in git ignore?

File: app/assets/stylesheets/main.scss [hunk 2]
```diff
@@ -1,105 +1,129 @@
...
+$success-color: #2fff2f;
+
 body {
   font-family: "Roboto Thin", serif
 }
 
-.bg {
-  background-color: purple;
-  color: white;
+.bg, .bg-primary {
```
Comment (usama-javaid-brainx) [inline app/assets/stylesheets/main.scss]: why do we have different classes here you can use one of them

File: app/assets/stylesheets/main.scss [hunk 3]
```diff
@@ -1,105 +1,129 @@
...
 }
+.success-messages {
+  text-align: left;
+  font-size: 12px;
+  padding: 5px;
+  color: $success-color;
+  font-style: italic;
+}
+
+.sticky-class {
```
Comment (usama-javaid-brainx) [inline app/assets/stylesheets/main.scss]: you can rename it to as position-sticky class

File: app/controllers/posts_controller.rb [hunk 4]
```diff
@@ -1,5 +1,45 @@
 class PostsController < ApplicationController
+  before_action :set_posts, only: [:edit, :update, :destroy]
```
Comment (usama-javaid-brainx) [inline app/controllers/posts_controller.rb]: naming convention issue you should name singular as set_post not set_posts

File: app/controllers/posts_controller.rb [hunk 5]
```diff
@@ -1,5 +1,45 @@
...
+      redirect_to root_path
+    else
+      render :edit, status: :unprocessable_entity
+    end
+  end
+
+  def destroy
+    @post.destroy!
+    redirect_to root_path
+  end
```
Comment (usama-javaid-brainx) [inline app/controllers/posts_controller.rb]: you are not showing notification messages here in this crud

File: app/controllers/reactions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  def like
+    if @reaction.blank?
+      @reaction = Reaction.create(user: current_user, post: @post, reaction_type: params[:reaction_type])
+      @reaction.save!
+    elsif @reaction.reaction_type == true
+      @reaction.update(reaction_type: nil)
+    else
+      @reaction.update(reaction_type: true)
+    end
+    redirect_to root_path
```
Comment (usama-javaid-brainx) [inline app/controllers/reactions_controller.rb]: need to update the if else here use find_or_initialize_by here

File: app/controllers/reactions_controller.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  def dislike
+    if @reaction.blank?
+      @reaction = Reaction.create(user: current_user, post: @post, reaction_type: params[:reaction_type])
+      @reaction.save!
+    elsif @reaction.reaction_type == false
+      @reaction.update(reaction_type: nil)
+    else
+      @reaction.update(reaction_type: false)
+    end
+    redirect_to root_path
```
Comment (usama-javaid-brainx) [inline app/controllers/reactions_controller.rb:25]: same here

File: app/controllers/reactions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  end
+
+  private
+
+  def set_variables
+    @user = User.find(params[:user_id])
```
Comment (usama-javaid-brainx) [inline app/controllers/reactions_controller.rb]: why you are not using current_user for the user who likes or dislike post?

File: app/controllers/reactions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+      @reaction.update(reaction_type: nil)
+    else
+      @reaction.update(reaction_type: false)
+    end
+    redirect_to root_path
+  end
+
+  private
+
+  def set_variables
```
Comment (usama-javaid-brainx) [inline app/controllers/reactions_controller.rb]: change name set_variable is too vague

File: app/models/friendship.rb [hunk 10]
```diff
@@ -1,14 +1,14 @@
 class Friendship < ApplicationRecord
```
Comment (usama-javaid-brainx) [inline app/models/friendship.rb]: friendship is not a valid name

File: app/models/friendship.rb [hunk 11]
```diff
@@ -1,14 +1,14 @@
...
   belongs_to :user
   belongs_to :friend, class_name: 'User'
 
   validates :user, presence: true
   validates :friend, presence: true, uniqueness: { scope: :user }
   validate :not_self
 
+  after_create :create_inverse_relationship
+  after_destroy :destroy_inverse_relationship
+
```
Comment (usama-javaid-brainx) [inline app/models/friendship.rb]: why you are not using self association for friends ?

File: app/models/post.rb [hunk 12]
```diff
@@ -1,2 +1,9 @@
 class Post < ApplicationRecord
+  belongs_to :user
+  has_one_attached :image
```
Comment (usama-javaid-brainx) [inline app/models/post.rb]: a post can have many images

File: app/models/reaction.rb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+class Reaction < ApplicationRecord
+  belongs_to :post
+  belongs_to :user
+
+  # validates :user, uniqueness: { scope: :post }
```
Comment (usama-javaid-brainx) [inline app/models/reaction.rb]: remove commented code

File: config/routes.rb [hunk 14]
```diff
@@ -1,22 +1,27 @@
...
-  resources :users, only: [:index, :show]
-
-  # resources :friend_requests
+  resources :friends
+  resources :users do
+    resources :posts do
+      resources :reactions do
+        collection do
+          get :like
+          get :dislike
```
Comment (usama-javaid-brainx) [inline config/routes.rb]: i dont think we need to make nested routes for reactions and post inside user

**Review by usama-javaid-brainx — CHANGES_REQUESTED** [review]

File: app/controllers/posts_controller.rb (hunk 5 above)
Comment (umair-khalid-brainx) [reply app/controllers/posts_controller.rb]: What notification messages ? Im not using notification messages anywhere.

File: app/controllers/reactions_controller.rb (hunk 6 above)
Comment (umair-khalid-brainx) [reply app/controllers/reactions_controller.rb]: find_or_initialize_by breaks my code logic that is why i did not use it

File: app/controllers/reactions_controller.rb (hunk 7 above)
Comment (umair-khalid-brainx) [reply app/controllers/reactions_controller.rb:25]: find_or_initialize_by breaks my code logic that is why i did not use it

File: app/models/friendship.rb (hunk 10 above)
Comment (umair-khalid-brainx) [reply app/models/friendship.rb]: Please suggest a valid name i think Friendship is appropriate

File: app/models/friendship.rb (hunk 11 above)
Comment (umair-khalid-brainx) [reply app/models/friendship.rb]: I dont understand how to implement a self association, we can discuss in person about this

File: app/controllers/posts_controller.rb (hunk 5 above)
Comment (usama-javaid-brainx) [reply app/controllers/posts_controller.rb]: its flash messages that you should show when an action is performed successfully just to give user a good experience

File: app/models/friendship.rb (hunk 11 above)
Comment (usama-javaid-brainx) [reply app/models/friendship.rb]: sure lets discuss

File: app/controllers/reactions_controller.rb (hunk 7 above)
Comment (umair-khalid-brainx) [reply app/controllers/reactions_controller.rb:25]: the logic you applied the other day is not working

File: app/controllers/reactions_controller.rb (hunk 6 above)
Comment (umair-khalid-brainx) [reply app/controllers/reactions_controller.rb]: the logic you applied the other day is not working

**Pull request merged**